internal/webhooks/clusterclass.go

```diff
@@ -154,17 +155,66 @@ func (webhook *ClusterClass) validate(ctx context.Context, old, new *clusterv1.C
 		}
 
 		// Ensure no MachineDeploymentClass currently in use has been removed from the ClusterClass.
-		allErrs = append(allErrs, webhook.validateRemovedMachineDeploymentClassesAreNotUsed(clusters, old, new)...)
+		allErrs = append(allErrs,
+			webhook.validateRemovedMachineDeploymentClassesAreNotUsed(clusters, old, new)...)
+
+		// Ensure no Variable would be invalidated by the update in spec
+		allErrs = append(allErrs,
+			webhook.validateVariablesNotInvalidated(clusters, old, new)...)
 	}
 	// Validate variables.
-	allErrs = append(allErrs, variables.ValidateClusterClassVariables(new.Spec.Variables, field.NewPath("spec", "variables"))...)
+	allErrs = append(allErrs,
+		variables.ValidateClusterClassVariables(new.Spec.Variables, field.NewPath("spec", "variables"))...,
+	)
 
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
 
+func (webhook *ClusterClass) validateVariablesNotInvalidated(clusters []clusterv1.Cluster, old, new *clusterv1.ClusterClass) field.ErrorList {
```

Comment:
I think this func has to work slightly differently:

* loop over all Variables in the ClusterClass and find out:
    * which variables have been removed
    * which variables have been changed (including required true/false & schema changes)
    * which variables have been added

Then we have to do the following:

* for removed variables:
    *  make sure they don't exist in any Cluster
* for changed variables:
    * if required: make sure the variable exist on all clusters and adheres to the schema
    * if not required: if the variable exist on a cluster it must adhere to the scheme, but it's okay if it doesn't exist
* for added variables:
    * if required: make sure the variable exist on all clusters and adheres to the schema
    * if not required: if the variable exist on a cluster it must adhere to the scheme, but it's okay if it doesn't exist
    
 By also handling "added variables" here we don't need `validateRequiredVariablesAreDefaulted` in `validateClusterClassVariable` anymore.

Reply:
Can be resolved.